### PR TITLE
[TRTLLM-7353][feat] Implement capturable drafting loops for speculation

### PR DESCRIPTION
**Pull request by mikeiovine**
## Description

Introduce a new concept called capturable drafting loops (CDLs) to the 2-model speculation infrastructure. The goal of this approach is to get the performance of 1-model speculation while keeping the modularity and flexibility of 2-model. This is achieved by doing the following:

1. Wrap the draft model and sampling loop in a sort of "meta module". 
2. This module can be injected into the `ModelEngine` and launched as a single CUDA graph. 

This is very much like the `SpecWorker` abstraction in the 1-model implementation, except it does not include verification. The advantages compared to standard 1-model are:

1. All speculation (ngram, draft/target, eagle) will be behind a unified `Drafter` abstraction. No more divergent code paths.
2. We can fall back to the old 2-model loop when certain features are not supported, e.g. xgrammar
3. We can create multiple CDLs that correspond to different draft lengths, different tree typologies, etc, and dispatch to the most performant one at runtime. We can also dynamically turn speculation off easily.
4. We keep the draft and target models and their KV cache manages completely decoupled, which improves overall flexibility. For example, different KV cache layouts and dtypes for draft and target models can be supported. You can also do stuff like running the draft and target on different GPUs.

To get the best performance, we need to finish overlap scheduling and also support CUDA graph for the first eagle3 iteration (which may include recomputation of KV cache for accepted tokens)

## Test Coverage

Covered by existing eagle3 2-model tests.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

* **New Features**
  * Optional capturable static drafting mode to produce multiple draft tokens in a single, CUDA-graph-friendly pass; lazy wrapper supports enabling this for draft models.

* **Performance**
  * Reduced iterative drafting overhead and improved latency; warmup and cache preparation now account for extra drafted tokens.

* **API Changes**
  * Model constructor exposes an optional drafting-wrapper parameter to enable wrapped/static drafting; LLM constructor gains an enable_autotuner flag in tests.

* **Compatibility**
  * Static drafting only enabled when guided decoding is off and mixed sampling is disabled.

### CODE REVIEW

**Review by yweng0828 — APPROVED** [review]

**Comment by mikeiovine** [comment]
> I have one last little question: IIUC, this PR is just a preliminary implementation. In the future, we may also replace sampler with TorchSampler and include features like reject sampling and so on. Is it correct?

Yes. There's a followup task to inject the sampler into the drafter.

File: tensorrt_llm/_torch/speculative/drafting_loops.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                new_draft_tokens.append(self.sample(logits))
+                new_position_ids += 1
+                attn_metadata.kv_lens_cuda[:batch_size] += 1
+                if i == 0:
+                    spec_metadata.hidden_states_read_indices[:batch_size].copy_(
+                        spec_metadata.hidden_states_write_indices[:batch_size])
+
+        return torch.stack(new_draft_tokens)
+
+    def sample(self, logits: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (mikeiovine) [reply tensorrt_llm/_torch/speculative/drafting_loops.py:112]: I will leave a TODO here

File: tensorrt_llm/_torch/pyexecutor/py_executor_creator.py [hunk 2]
```diff
@@ -252,13 +252,31 @@ def create_py_executor(
...
             draft_spec_config.max_draft_len = 0
 
+            use_chain_drafter = (
+                executor_config.guided_decoding_config is None
+                and not pytorch_backend_config.enable_mixed_sampler)
+
+            if use_chain_drafter:
+
+                # Make sure we allocate enough KV cache space.
+                spec_config._num_extra_kv_tokens = spec_config.max_draft_len - 1
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/py_executor_creator.py:266]: Note: this was an attempt to fix a bug where not enough KV cache was being allocated during CUDA graph warmup. It's not the right way to do this though; will revert it tomorrow.

**Review by IzzyPutterman — APPROVED** [review]

**Review by Funatiq — APPROVED** [review]

**Review by brb-nv — APPROVED** [review]
LGTM.

File: tensorrt_llm/_torch/speculative/model_drafter.py [hunk 3]
```diff
@@ -365,8 +375,29 @@ def prepare_draft_tokens(
                 for req in scheduled_requests.all_requests()
             }
 
-            # Initial forward pass
+            # Initial forward pass. May do the complete drafting loop
+            # if use_static_draft_loop is set.
             outputs = self._forward_draft_model(draft_batch, resource_manager)
+
+            if self.use_static_draft_loop:
+                outputs_host = outputs.cpu()
```
[the remark this answers is not in the document]
Comment (syuoni) [reply tensorrt_llm/_torch/speculative/model_drafter.py:383]: Oh I see, it's sync device tokens to cpu. To make it async, we need overlap scheduling. Closing this thread.

File: tensorrt_llm/_torch/speculative/model_drafter.py [hunk 4]
```diff
@@ -365,8 +375,29 @@ def prepare_draft_tokens(
...
+                            continue
+
+                        target_req = req_id_to_old_request[req.py_request_id]
+                        target_req.py_draft_tokens.append(
+                            outputs_host[token_idx][req_idx])
+
+                for req in draft_batch.all_requests():
+                    self.draft_seq_slot_manager.free_resources(req)
+
+                return
```
[the remark this answers is not in the document]
Comment (mikeiovine) [reply tensorrt_llm/_torch/speculative/model_drafter.py:399]: This can be done in a follow up
